service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+		DomainName string
+		// To query the target workflows for processing
+		QueryCondition string
+		// Reaason for the operation
+		Reason string
+		// Supporting: reset,terminate
+		BatchType string
+
+		// Below are all optional
+
+		// RPS of processing. Default to defaultRPS
+		// TODO we will implement smarter way than this static rate limiter: https://github.com/uber/cadence/issues/2138
+		RPS int
+		// Number of goroutines running in parallel to process
+		Concurrency int
+		// Number of attempts for each workflow to process in case of retryable error before giving up
+		AttemptsOnRetryableError int
+		ActivityHeartBeatTimeout time.Duration
+	}
+
+	HeartBeatDetails struct {
+		PageToken   []byte
+		CurrentPage int
+		// This is just an estimation for visibility
+		TotalEstimate int64
+		// Number of workflows processed successfully
+		SuccessCount int32
+		// Number of workflows that give up due to errors.
+		ErrorCount int32
+	}
+
+	taskDetail struct {
+		execution shared.WorkflowExecution
+		attempts  int
+		// passing along the current heartbeat details to make heartbeat within a task so that it won't timeout
+		hbd HeartBeatDetails
+	}
+)
+
+var (
+	batchActivityRetryPolicy = cadence.RetryPolicy{
+		InitialInterval:    10 * time.Second,
+		BackoffCoefficient: 1.7,
+		MaximumInterval:    5 * time.Minute,
+		ExpirationInterval: infiniteDuration,
+	}
+
+	batchActivityOptions = workflow.ActivityOptions{
+		ScheduleToStartTimeout: 5 * time.Minute,
+		StartToCloseTimeout:    infiniteDuration,
+		RetryPolicy:            &batchActivityRetryPolicy,
+	}
+)
+
+func init() {
+	workflow.RegisterWithOptions(BatchWorkflow, workflow.RegisterOptions{Name: batchWFTypeName})
+	activity.RegisterWithOptions(BatchActivity, activity.RegisterOptions{Name: batchActivityName})
+}
+
+// BatchWorkflow is the workflow that runs a batch job of resetting workflows
+func BatchWorkflow(ctx workflow.Context, batchParams BatchParams) error {
+	batchParams, err := setDefaultParams(batchParams)
+	if err != nil {
+		return err
+	}
+	batchActivityOptions.HeartbeatTimeout = batchParams.ActivityHeartBeatTimeout
+	opt := workflow.WithActivityOptions(ctx, batchActivityOptions)
+	return workflow.ExecuteActivity(opt, batchActivityName, batchParams).Get(ctx, nil)
+}
+
+func setDefaultParams(params BatchParams) (BatchParams, error) {
+	if params.BatchType == "" || params.Reason == "" || params.DomainName == "" || params.QueryCondition == "" {
+		return BatchParams{}, fmt.Errorf("must provide required parameters: BatchType/Reason/DomainName/QueryCondition")
+	}
+	if params.RPS <= 0 {
+		params.RPS = defaultRPS
+	}
+	if params.Concurrency <= 0 {
+		params.Concurrency = defaultConcurrency
+	}
+	if params.AttemptsOnRetryableError <= 0 {
+		params.AttemptsOnRetryableError = defaultAttemptsOnRetryableError
+	}
+	if params.ActivityHeartBeatTimeout <= 0 {
+		params.ActivityHeartBeatTimeout = defaultActivityHeartBeatTimeout
+	}
+	switch params.BatchType {
+	case BatchTypeReset:
```

Comment:
I feel this switch could be simpler, could we just:
```
case Terminate:
  // do something or nothing
default:
  // not support
```
and later for each support type we could add a case.

(The current usage of `fallthrough` makes it a little bit harder to read IMHO)